drivers/include/tla2528.hpp

```diff
@@ -0,0 +1,135 @@
+#pragma once
+#include <libhal/i2c.hpp>
+#include <libhal/steady_clock.hpp>
+#include <libhal/units.hpp>
+
+namespace sjsu::drivers {
+
+// adapters
+class tla2528_adc;
+class tla2528_input_pin;
+class tla2528_output_pin;
+
+/**
+ * @brief tla2528 gpio expander & adc driver
+ *
+ * tla2528 has 8 pins which can be independently operated as an adc,
+ * digital input, and digital out over i2c. The i2c address is configured by
+ * resistors on the chip There are no options for internal pull up or pull down
+ * resistors. The output pins have the option of push-pull or open-drain. When
+ * in adc mode there is an option to increase reading granularity though
+ * sampling averaging (unimplemented).
+ */
+class tla2528
+{
+public:
+  enum class pin_mode : hal::byte
+  {
+    analog_input,
+    digital_input,
+    digital_output_open_drain,
+    digital_output_push_pull
+  };
+
+  tla2528(hal::i2c& p_i2c, hal::byte p_i2c_address = default_address);
+  /**
+   * @brief set what service a pin will provide
+   */
+  void set_pin_mode(pin_mode p_mode, hal::byte p_channel);
+
+  // If channel is not set to correct config then undefined behavior will occur
```

Comment:
UB is good for performance but we shouldn't invoke it if we aren't getting any benefit from it. We should define the undefined.

Reply:
Could you clarify what you mean by that? Do you mean to have safety checks to prevent undefined calls, document the undefined behavior, or something else?

Reply:
I think I understand what you meant now from another conversation. I updated the documentation, look it over.